# LAB 01 - LÀM QUEN VỚI PYTORCH

**Họ và tên:** ............................................................  
**MSSV:** .................................................................  

Notebook thực hiện các bài 1-11 theo file yêu cầu. Một số seed được đặt để kết quả ngẫu nhiên có thể chạy lại giống nhau.

## Bài 1. Tạo Tensor cơ bản

**Yêu cầu:** Tạo Tensor số nguyên kích thước 2 x 3; cho biết kích thước, số chiều và kiểu dữ liệu.

In [1]:
import torch

a = torch.tensor(
    [[1, 2, 3],
     [4, 5, 6]],
    dtype=torch.int32
)

print(a)
print("Shape:", a.shape)
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)


tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
Shape: torch.Size([2, 3])
Number of dimensions: 2
Data type: torch.int32


### Trả lời / nhận xét

- Tensor `a` có kích thước `torch.Size([2, 3])`, nghĩa là 2 hàng và 3 cột.
- Số chiều của Tensor là `2`.
- Kiểu dữ liệu là `torch.int32`.

## Bài 2. Tạo các Tensor thường dùng

**Yêu cầu:** Tạo Tensor ngẫu nhiên, Tensor toàn 0, toàn 1, ma trận đơn vị; nhận xét torch.rand() và torch.randn(); tạo Tensor 5 x 5 toàn giá trị 10.

In [2]:
import torch

torch.manual_seed(42)
torch.set_printoptions(precision=4, sci_mode=False)

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
f = torch.full((5, 5), 10)

print("torch.rand(3, 4):\n", a)
print("\ntorch.randn(3, 4):\n", b)
print("\nzeros:\n", c)
print("\nones:\n", d)
print("\neye(4):\n", e)
print("\nTensor 5 x 5 toàn giá trị 10:\n", f)


torch.rand(3, 4):
 tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])

torch.randn(3, 4):
 tensor([[ 2.2082, -0.6380,  0.4617,  0.2674],
        [ 0.5349,  0.8094,  1.1103, -1.6898],
        [-0.9890,  0.9580,  1.3221,  0.8172]])

zeros:
 tensor([[0., 0., 0.],
        [0., 0., 0.]])

ones:
 tensor([[1., 1., 1.],
        [1., 1., 1.]])

eye(4):
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])

Tensor 5 x 5 toàn giá trị 10:
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


### Trả lời / nhận xét

- `torch.rand()` sinh số thực ngẫu nhiên theo phân phối đều trong khoảng `[0, 1)`.
- `torch.randn()` sinh số thực ngẫu nhiên theo phân phối chuẩn có trung bình xấp xỉ 0 và độ lệch chuẩn xấp xỉ 1, vì vậy có thể xuất hiện cả số âm và số dương.
- Tensor 5 x 5 toàn giá trị 10 có thể tạo bằng `torch.full((5, 5), 10)`.

## Bài 3. Kiểm tra thông tin Tensor

**Yêu cầu:** Dùng shape, size(), ndim và dtype; giải thích ý nghĩa của torch.Size([2, 3, 4]).

In [3]:
import torch

a = torch.zeros(2, 3, 4)

print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)


Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


### Trả lời / nhận xét

`torch.Size([2, 3, 4])` cho biết Tensor có 3 chiều. Kích thước lần lượt của ba chiều là 2, 3 và 4. Tổng số phần tử là `2 x 3 x 4 = 24`.

## Bài 4. Slicing và Indexing

**Yêu cầu:** Cho biết shape của a[1] và a[1:, 2:4]; lấy hai phần tử cuối cùng trên chiều thứ ba.

In [4]:
import torch

torch.manual_seed(7)
a = torch.randn(3, 4, 5)

print("Original tensor:")
print(a)
print("\nTensor a[1]:")
print(a[1])
print("Shape a[1]:", a[1].shape)
print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])
print("Shape a[1:, 2:4]:", a[1:, 2:4].shape)
print("\nHai phần tử cuối trên chiều thứ ba - a[:, :, -2:]:")
print(a[:, :, -2:])
print("Shape:", a[:, :, -2:].shape)


Original tensor:
tensor([[[-0.8201,  0.3956,  0.8989, -1.3884, -0.1670],
         [ 0.2851, -0.6411, -0.8937,  0.9265, -0.5355],
         [-1.1597, -0.4602,  0.7085,  1.0128,  0.2304],
         [ 1.0902, -1.5827, -0.3246,  1.9264, -0.3300]],

        [[ 0.1984,  0.7821,  1.0391, -0.7245, -0.2093],
         [-0.2153, -1.8157, -0.3452, -2.0615,  0.6741],
         [-1.3233, -1.3598, -0.0835, -0.0235,  0.1744],
         [ 2.2983,  0.9571, -0.6619, -0.8285, -0.6057]],

        [[-1.4013,  1.2973,  1.6409, -1.0567, -0.5321],
         [ 0.2991, -1.4204, -0.1635,  0.1332,  0.1825],
         [-0.5146,  0.8005, -0.9388, -1.2370, -0.0626],
         [ 0.4495, -0.8858,  1.1970,  0.4675, -0.1439]]])

Tensor a[1]:
tensor([[ 0.1984,  0.7821,  1.0391, -0.7245, -0.2093],
        [-0.2153, -1.8157, -0.3452, -2.0615,  0.6741],
        [-1.3233, -1.3598, -0.0835, -0.0235,  0.1744],
        [ 2.2983,  0.9571, -0.6619, -0.8285, -0.6057]])
Shape a[1]: torch.Size([4, 5])

Tensor a[1:, 2:4]:
tensor([[[-1.3233, 

### Trả lời / nhận xét

- `a[1]` chọn phần tử có chỉ số 1 ở chiều đầu tiên, nên bỏ chiều đầu và còn shape `(4, 5)`.
- `a[1:, 2:4]` giữ 2 phần tử ở chiều đầu (`1, 2`), chọn 2 phần tử ở chiều thứ hai (`2, 3`) và giữ nguyên chiều thứ ba, nên shape là `(2, 2, 5)`.
- Hai phần tử cuối cùng của chiều thứ ba được lấy bằng `a[:, :, -2:]`, có shape `(3, 4, 2)`.

## Bài 5. Reshape và Flatten

**Yêu cầu:** Kiểm tra số phần tử trước/sau reshape; thử reshape thành kích thước 2 x 6.

In [5]:
import torch

torch.manual_seed(11)
a = torch.randn(3, 4)

print("Original:")
print(a)
print("Số phần tử ban đầu:", a.numel())

flat = a.ravel()
print("\nFlatten:")
print(flat)
print("Số phần tử sau flatten:", flat.numel())

r1 = a.reshape(3, 2, 2)
print("\nReshape to 3 x 2 x 2:")
print(r1)
print("Số phần tử sau reshape 3 x 2 x 2:", r1.numel())

r2 = a.reshape(2, 6)
print("\nReshape to 2 x 6:")
print(r2)
print("Số phần tử sau reshape 2 x 6:", r2.numel())


Original:
tensor([[ 0.7376,  1.9459, -0.6995, -1.3023],
        [-0.5133, -0.2696,  0.2462,  0.4839],
        [ 0.4504, -0.9568,  1.5012, -0.3136]])
Số phần tử ban đầu: 12

Flatten:
tensor([ 0.7376,  1.9459, -0.6995, -1.3023, -0.5133, -0.2696,  0.2462,  0.4839,
         0.4504, -0.9568,  1.5012, -0.3136])
Số phần tử sau flatten: 12

Reshape to 3 x 2 x 2:
tensor([[[ 0.7376,  1.9459],
         [-0.6995, -1.3023]],

        [[-0.5133, -0.2696],
         [ 0.2462,  0.4839]],

        [[ 0.4504, -0.9568],
         [ 1.5012, -0.3136]]])
Số phần tử sau reshape 3 x 2 x 2: 12

Reshape to 2 x 6:
tensor([[ 0.7376,  1.9459, -0.6995, -1.3023, -0.5133, -0.2696],
        [ 0.2462,  0.4839,  0.4504, -0.9568,  1.5012, -0.3136]])
Số phần tử sau reshape 2 x 6: 12


### Trả lời / nhận xét

Tensor ban đầu có 12 phần tử. Cả `3 x 2 x 2` và `2 x 6` đều có 12 phần tử, nên reshape thực hiện được và không làm thay đổi dữ liệu, chỉ thay đổi cách tổ chức các phần tử.

## Bài 6. Phép toán và nhân ma trận

**Yêu cầu:** Phân biệt a * b với a @ b.T; cho biết shape của kết quả a @ b.T.

In [6]:
import torch

torch.manual_seed(21)
a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a:")
print(a)
print("b:")
print(b)
print("\na + b:")
print(a + b)
print("\na / b:")
print(a / b)
print("\na**2:")
print(a ** 2)
print("\nElement-wise multiplication a * b:")
print(a * b)
print("Shape a * b:", (a * b).shape)
print("\nMatrix multiplication a @ b.T:")
matmul_result = a @ b.T
print(matmul_result)
print("Shape a @ b.T:", matmul_result.shape)


a:
tensor([[ 0.1081, -0.4376, -0.7697],
        [-0.1929, -0.3626, -2.8451]])
b:
tensor([[ 1.4435,  0.4976,  0.6542],
        [ 0.0754, -1.0767,  0.1269]])

a + b:
tensor([[ 1.5516,  0.0600, -0.1155],
        [-0.1175, -1.4393, -2.7182]])

a / b:
tensor([[  0.0749,  -0.8794,  -1.1765],
        [ -2.5573,   0.3367, -22.4128]])

a**2:
tensor([[0.0117, 0.1915, 0.5924],
        [0.0372, 0.1315, 8.0946]])

Element-wise multiplication a * b:
tensor([[ 0.1560, -0.2177, -0.5035],
        [-0.0146,  0.3904, -0.3612]])
Shape a * b: torch.Size([2, 3])

Matrix multiplication a @ b.T:
tensor([[-0.5652,  0.3816],
        [-2.3202,  0.0147]])
Shape a @ b.T: torch.Size([2, 2])


### Trả lời / nhận xét

- `a * b` là phép nhân từng phần tử tương ứng, vì `a` và `b` cùng shape `(2, 3)` nên kết quả cũng có shape `(2, 3)`.
- `a @ b.T` là phép nhân ma trận. `a` có shape `(2, 3)`, `b.T` có shape `(3, 2)`, nên kết quả có shape `(2, 2)`.

## Bài 7. Các hàm thống kê trên Tensor

**Yêu cầu:** Thay dim=0 bằng dim=1 và nhận xét sự khác nhau.

In [7]:
import torch

torch.manual_seed(31)
a = torch.randn(3, 4)
print("a:")
print(a)

print("\n--- dim = 0 (theo cột) ---")
print("Mean:", torch.mean(a, dim=0))
print("Standard deviation:", torch.std(a, dim=0))
print("Cumulative sum:\n", torch.cumsum(a, dim=0))

print("\n--- dim = 1 (theo hàng) ---")
print("Mean:", torch.mean(a, dim=1))
print("Standard deviation:", torch.std(a, dim=1))
print("Cumulative sum:\n", torch.cumsum(a, dim=1))


a:
tensor([[ 0.3519, -0.5683, -0.3886, -0.2856],
        [-0.4513,  0.9157, -0.7624,  1.0557],
        [-0.1154,  0.5772,  0.4651,  0.2977]])

--- dim = 0 (theo cột) ---
Mean: tensor([-0.0716,  0.3082, -0.2286,  0.3559])
Standard deviation: tensor([0.4034, 0.7777, 0.6292, 0.6725])
Cumulative sum:
 tensor([[ 0.3519, -0.5683, -0.3886, -0.2856],
        [-0.0994,  0.3474, -1.1510,  0.7701],
        [-0.2148,  0.9246, -0.6859,  1.0678]])

--- dim = 1 (theo hàng) ---
Mean: tensor([-0.2226,  0.1894,  0.3061])
Standard deviation: tensor([0.4004, 0.9299, 0.3036])
Cumulative sum:
 tensor([[ 0.3519, -0.2164, -0.6050, -0.8906],
        [-0.4513,  0.4644, -0.2980,  0.7577],
        [-0.1154,  0.4618,  0.9269,  1.2245]])


### Trả lời / nhận xét

- Với `dim=0`, PyTorch gộp theo chiều hàng để tính cho từng cột; vì Tensor có 4 cột nên mean/std trả về 4 giá trị.
- Với `dim=1`, PyTorch gộp theo chiều cột để tính cho từng hàng; vì Tensor có 3 hàng nên mean/std trả về 3 giá trị.
- `cumsum(dim=0)` cộng tích lũy theo chiều dọc, còn `cumsum(dim=1)` cộng tích lũy theo chiều ngang.

## Bài 8. Autograd cơ bản

**Yêu cầu:** Tính thủ công đạo hàm y=x^2 tại x=3.6; đổi thành y=3x^2+2x+1 và kiểm tra gradient.

In [8]:
import torch

x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()
print("Trường hợp 1: y = x^2")
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)
print("Đạo hàm thủ công = 2*x =", 2 * 3.6)

x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * x2**2 + 2 * x2 + 1
y2.backward()
print("\nTrường hợp 2: y = 3*x^2 + 2*x + 1")
print("y =", y2)
print("x.grad =", x2.grad)
print("Đạo hàm thủ công = 6*x + 2 =", 6 * 3.6 + 2)


Trường hợp 1: y = x^2
x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)
Đạo hàm thủ công = 2*x = 7.2

Trường hợp 2: y = 3*x^2 + 2*x + 1
y = tensor(47.0800, grad_fn=<AddBackward0>)
x.grad = tensor(23.6000)
Đạo hàm thủ công = 6*x + 2 = 23.6


### Trả lời / nhận xét

- Với `y = x^2`, đạo hàm là `dy/dx = 2x`; tại `x = 3.6` ta được `7.2`, trùng với `x.grad`.
- Với `y = 3x^2 + 2x + 1`, đạo hàm là `dy/dx = 6x + 2`; tại `x = 3.6` ta được `23.6`, trùng với gradient do autograd tính.

## Bài 9. Gradient Descent tìm hệ số đa thức

**Yêu cầu:** Sinh dữ liệu từ y=x^2+2x+3, dùng NAdam và autograd tìm lại ba hệ số; giải thích zero_grad(), backward(), step().

In [9]:
import numpy as np
import torch

# Đặt seed để kết quả có thể chạy lại giống nhau.
np.random.seed(42)
torch.manual_seed(25)

polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)
print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("\nLearned coefficients:")
print(w)
print("Final MSE:", mse.item())
print("Target coefficients: [1, 2, 3]")


Initial coefficients:
tensor([[1.1781],
        [0.1056],
        [1.5313]], requires_grad=True)



Learned coefficients:
tensor([[1.0000],
        [2.0000],
        [2.9999]], requires_grad=True)
Final MSE: 8.294494868721358e-09
Target coefficients: [1, 2, 3]


### Trả lời / nhận xét

Ba hệ số học được xấp xỉ `[1, 2, 3]`, phù hợp với đa thức dùng để sinh dữ liệu.

- `optimizer.zero_grad()`: xóa gradient của lần lặp trước vì PyTorch mặc định cộng dồn gradient.
- `mse.backward()`: thực hiện lan truyền ngược để tính gradient của hàm mất mát theo các tham số `w`.
- `optimizer.step()`: dùng gradient vừa tính để cập nhật các hệ số `w` theo thuật toán NAdam.

## Bài 10. Giải hệ phương trình bằng Autograd

**Yêu cầu:** Kiểm tra nghiệm với A+B=9, C-D=1, A+C=8, B-D=2; nhận xét vì sao có thể có nhiều nghiệm.

In [10]:
import random
import torch

random.seed(42)
A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4

    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)
print("\nKiểm tra bốn phương trình:")
print("A + B =", (A + B).item(), "~ 9")
print("C - D =", (C - D).item(), "~ 1")
print("A + C =", (A + C).item(), "~ 8")
print("B - D =", (B - D).item(), "~ 2")
print("Tổng sai số bình phương =", sqerr.item())


A = tensor(4.7772, requires_grad=True)
B = tensor(4.2227, requires_grad=True)
C = tensor(3.2228, requires_grad=True)
D = tensor(2.2227, requires_grad=True)

Kiểm tra bốn phương trình:
A + B = 8.999945640563965 ~ 9
C - D = 1.0000205039978027 ~ 1
A + C = 7.999980926513672 ~ 8
B - D = 1.9999852180480957 ~ 2
Tổng sai số bình phương = 3.957666194764897e-09


### Trả lời / nhận xét

Nghiệm thu được thỏa bốn phương trình với sai số rất nhỏ.

Hệ có nhiều nghiệm vì bốn phương trình không độc lập hoàn toàn. Đặt `A = t` thì:
`B = 9 - t`, `C = 8 - t`, `D = 7 - t`. Vì `t` có thể nhận nhiều giá trị nên tồn tại vô số nghiệm. Thuật toán tối ưu sẽ hội tụ tới một nghiệm cụ thể tùy giá trị khởi tạo.

## Bài 11. Bài tập áp dụng

**Yêu cầu:** (1) Tensor X 24 phần tử reshape 4x6, 2x12, 2x3x4. (2) Hai Tensor A, B kích thước 3x4 và các phép toán/thống kê. (3) Autograd cho y=2x^2+5x+3 tại x=2. (4) Huấn luyện đa thức y=2x^2-3x+5.

In [11]:
import numpy as np
import torch

# 1) Tensor X có 24 phần tử và reshape
X = torch.arange(24)
print("1) Tensor X:")
print(X)
print("\nX reshape 4 x 6:\n", X.reshape(4, 6))
print("\nX reshape 2 x 12:\n", X.reshape(2, 12))
print("\nX reshape 2 x 3 x 4:\n", X.reshape(2, 3, 4))

# 2) Hai Tensor A, B kích thước 3 x 4
torch.manual_seed(101)
A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("\n2) A:\n", A)
print("B:\n", B)
print("A + B:\n", A + B)
print("A * B:\n", A * B)
print("A / B:\n", A / B)
print("Trung bình theo cột của A:", torch.mean(A, dim=0))
print("Độ lệch chuẩn theo cột của A:", torch.std(A, dim=0))
print("Trung bình theo cột của B:", torch.mean(B, dim=0))
print("Độ lệch chuẩn theo cột của B:", torch.std(B, dim=0))

# 3) Autograd: y = 2x^2 + 5x + 3 tại x = 2
x_auto = torch.tensor(2.0, requires_grad=True)
y_auto = 2*x_auto**2 + 5*x_auto + 3
y_auto.backward()
print("\n3) y = 2x^2 + 5x + 3 tại x = 2")
print("y =", y_auto.item())
print("Autograd dy/dx =", x_auto.grad.item())
print("Thủ công dy/dx = 4x + 5 =", 4*2 + 5)

# 4) Huấn luyện đa thức y = 2x^2 - 3x + 5
np.random.seed(42)
torch.manual_seed(20)
polynomial = np.poly1d([2, -3, 5])
N = 20
X_np = np.random.randn(N, 1) * 5
Y_np = polynomial(X_np)
XX = np.hstack([X_np*X_np, X_np, np.ones_like(X_np)])

w = torch.randn(3, 1, requires_grad=True)
x_train = torch.tensor(XX, dtype=torch.float32)
y_train = torch.tensor(Y_np, dtype=torch.float32)
optimizer = torch.optim.NAdam([w], lr=0.01)

# Tăng số epoch để các hệ số hội tụ rõ ràng hơn trong phần bài tập áp dụng.
for epoch in range(5000):
    optimizer.zero_grad()
    y_pred = x_train @ w
    mse = torch.mean((y_train - y_pred)**2)
    mse.backward()
    optimizer.step()

print("\n4) Learned coefficients for y = 2x^2 - 3x + 5:")
print(w)
print("Final MSE:", mse.item())
print("Target coefficients: [2, -3, 5]")


1) Tensor X:
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])

X reshape 4 x 6:
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

X reshape 2 x 12:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

X reshape 2 x 3 x 4:
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

2) A:
 tensor([[-1.3905, -0.8152, -0.3204,  0.7377],
        [-1.7534,  0.6033, -0.2520, -0.4373],
        [-0.5728, -3.2023,  0.5633,  0.2153]])
B:
 tensor([[-2.1962, -0.2166, -0.7262,  1.8464],
        [ 0.2487,  0.2779,  1.6179, -0.2337],
        [-1.3387, -0.0620, -0.2693,  0.8990]])
A + B:
 tensor([[-3.5867, -1.0319, -1.0466,  2.5842],
        [-1.5047,  0.8812,  1.3659, -0.6710],
  


4) Learned coefficients for y = 2x^2 - 3x + 5:
tensor([[ 2.0002],
        [-2.9999],
        [ 4.9891]], requires_grad=True)
Final MSE: 7.071181607898325e-05
Target coefficients: [2, -3, 5]


### Trả lời / nhận xét

- Các phép reshape của `X` đều hợp lệ vì tổng số phần tử luôn bằng 24.
- Các phép `A+B`, `A*B`, `A/B` là phép toán theo phần tử; mean/std với `dim=0` cho kết quả theo từng cột.
- Với `y = 2x^2 + 5x + 3`, đạo hàm là `4x + 5`; tại `x=2` bằng `13`, khớp với autograd.
- Sau huấn luyện, ba hệ số xấp xỉ `[2, -3, 5]`, đúng với đa thức sinh dữ liệu. Ở phần áp dụng tăng số epoch để minh họa sự hội tụ rõ hơn.